### PR TITLE
feat: auto-recover from pruned node errors during extraction

### PR DESCRIPTION
**Pull request by Cordtus**
## Summary

Simplifies pruned node handling by probing at startup instead of reactive recovery during extraction.

## Changes

- Probe earliest available block height during startup
- Adjusts start height to match earliest available height

## Details

Since there is no current gRPC method that can access the required value directly until [cosmos/cosmos-sdk #25647](https://github.com/cosmos/cosmos-sdk/pull/25647) is merged.
In the meantime, we can fetch that same value by forcing a particular error as a silly workaround:

1. If DB has data, resume from `latest indexed + 1` (no probe)
2. If DB is empty, attempt to fetch block height `1`, which if not exists will return the earliest available height
3. If response returns "lowest height is X" error, start indexing from height X

## Test

Unit test for error message parsing in `internal/utils/block_test.go`.

Tested manually using a pruned Juno node - behaves as expected.

### CODE REVIEW

**Review by fmorency — CHANGES_REQUESTED** [review]
Thanks @Cordtus!

As discussed on TG, you could do it in multiple stages.

1. Try extracting at block 1. If it works, start the full extraction routine from block 1.
2. If it fails, detect the block from the error message (N)
3. Try extracting at block N. If it works, start the full extraction routine from block N
4. If it fails, detect the block from the error message again (N'). If N' == N, exit with error.
5. Try extracting at block N'. If it works, start the full extraction routine from block N'. N = N'

Repeat  3-5. You could even add some heuristic if it fails too many time, like N' = N + 100.

E.g.
```
> Can I extract from block 1?
No, the lowest is block 12345
> Can I extract from block 12345?
No, the lowest is block 15555 (we hit another node, that happens)
> Can I extract from block 15555?
No, the lowest is block 23456 (we hit yet another node, that happens)
> Can I extract from block 23456?
No, the lowest is block 23666 (we hit yet another node, that happens)
> Aight, can I extract from block 23766 (heuristic N + 100)
Yes, that works, let's go!
```

Phase 1 is "detect what the minimum block is". It *can* fail even if you get the minimum block from the error message, because we're never sure which node we'll hit, and they could all be at a different height, thus repeat max M times and/or use a heuristic if we can't determine the minimum block after 3 tries.

You wouldn't need the new signal at all and should be able to re-use existing code without (much) modifications.

Phase 2 is the full extraction loop.

You could reuse the approach above once Cosmos merges https://github.com/cosmos/cosmos-sdk/pull/25647 and https://github.com/cosmos/cosmos-sdk/pull/25648. Instead of parsing the error message to find the minimum block, you'd poke the gRPC endpoint. Same retry logic.

WDYT?

**Comment by Cordtus** [comment]
Reworked this one, and updated the description accordingly.

File: internal/utils/block.go [hunk 1]
```diff
@@ -25,3 +29,46 @@ func GetLatestBlockHeightWithRetry(gRPCClient *client.GRPCClient, maxRetries uin
 		},
 	)
 }
+
+// GetEarliestBlockHeight determines the earliest available block on a node.
+// It probes block 1 to check if the node is an archive node or pruned.
+// For archive nodes, returns 1. For pruned nodes, parses the error message
+// to extract the lowest available height.
+func GetEarliestBlockHeight(gRPCClient *client.GRPCClient, maxRetries uint) (uint64, error) {
```
Comment (fmorency) [inline internal/utils/block.go:48]: Can we make this function a little more robust like suggested in the first review?

I saw cases there the lowest height from error wasn't working because the query hit another node and it didn't have that height. I.e., the other node has a lowest height higher than previously reported.

**Review by fmorency — CHANGES_REQUESTED** [review]
Thank you! Much better! A single minor comment.

nit: Add more tests but we can do that in another PR. Would love e2e tests to cover some edge-cases like the ones I described in my other review.

File: internal/utils/block.go (hunk 1 above)
Comment (Cordtus) [reply internal/utils/block.go:48]: Ah I didn't consider the case of load balancers using nodes with varying heights.. is this common?

Comment (fmorency) [reply internal/utils/block.go:48]: > Ah I didn't consider the case of load balancers using nodes with varying heights.. is this common?

I encountered this issue multiple times while building this project, primarily with Osmosis and the Hub. I'm not sure if it's common, but I believe it's common enough to address. I'm surprised you didn't encounter this issue during your tests.

Comment (Cordtus) [reply internal/utils/block.go:48]: I typically use my own nodes for dev/testing (mainly because of issues like this with public ones).
I will have something for this early tomorrow.

Comment (Cordtus) [reply internal/utils/block.go:48]: @fmorency how do you feel about a runtime flag for load balancers?
Somethihg like...

- Does db exist?
 - YES - find latest height
 - NO - attempt starting from height 1
 
- Does fetching this height result in the handy error message?
 - YES - start fetching from this "laterEarliest" heght
 - NO - start fetching normally
 
- Are there further missing block errors?
 - YES - Assuming the given height is  > current "laterEarliest" height, replace it with this one and fall back to previous step, retrying tihs loop up to X times. After X retries, either try something like 'start indexing from ("laterEarliest"+1000) height' or just error out / panic


```mermaid
flowchart TD
    A[Start] --> B{DB exists?}
    B -->|Yes| C[Find latest height]
    B -->|No| D[Set height = 1]
    C --> E[Fetch height]
    D --> E
    E --> F{laterEarliest error?}
    F -->|No| G[Continue indexing]
    F -->|Yes| H{New laterEarliest > current?}
    H -->|No| I[Error/Panic]
    H -->|Yes| J[Update height = new laterEarliest]
    J --> K{Retries < X?}
    K -->|Yes| L[retries++]
    L --> E
    K -->|No| M["height = laterEarliest + offset (default: 1000)"]
    M --> N[Fetch from height]
    N --> O{Success?}
    O -->|Yes| G
    O -->|No| I
```

    style Q fill:#f9f,stroke:#333
    note[offset default: 1000, configurable via flag]
